[MIG][10.0] medical

lasley opened the pull request:
Done:
- Bump version
- Change openerp namespace to odoo
- Fix inverse on res_partner to patient

Left to do:
- [x] Remove all is_\* fields (e.g. is_school, is_work, etc.) from res.partner and move to res.partner.type.
- [x] Remove medical_constants
- [x] Remove use of `use_parent_address`

It would be nice if also:
- [x] Existing module is renamed to `medical_patient`. _(future)_
- [x] `medical_patient` The views for entities is repeated in a lot of spots (patient, physician, pharmacy, etc). We need abstract views to inherit so that we can reduce a lot of the code duplication. Something like the view provided in OCA/web#490 just not for stages
- [x] Module `medical` is nothing but an install wizard to direct installation of appropriate `medical_*` modules & the base `res.partner.type` logic  _(future)_
- [x] Tours were a thing
- [x]  Menu redesign #127  _(future)_

lasley committed:
[9.0][MIG] medical (#69)

* [MIG] Upgrade medical to 9.0.
* [FIX] medical: Fix file headers, add social insurance numbers to patients plus some other small additions and changes.
* [IMP] medical: Refactor tests to use demo data, add api.depends to some medical_patient model compute methods.
* [FIX] medical: Various adjustments to field definitions, patient logic. Add items to roadmap in readme.
* [FIX] medical: Fix help descriptions in res.partner file.
* [FIX] medical: Remove string attributes where the name of the field is the same as the string in res.partner.

lasley committed:
[MIG] medical: Upgrade medical to v10
(+20 squashed commits)
Squashed commits:
[16b317d] [FIX] medical: Gary Smith email
* Fix incorrect email for Gary Smith patient demo
[9877361] [FIX] medical: Inherit mail thread for medical entity
* Fixes col missing error for `medical_patient.message_last_post`
[d3c1903] Switch ID computation to char & add inverse
[e3b4a68] Add documentation on abstract entity to readme and fix a few tests
[52ca175] PR fixes
[548994c] Add pass thru method
[45f3f41] Add tests for new methods
[d1aef73] [IMP] medical: Add default image and attributes
* Add default image handling for medical entities in the model
* Add more attributes to the views to allow slightly easier inheritance
* Add mode=primary to patient views
[f09f733] Internal PR review fixes
[575a62d] Add patient avatar and more image targets for xpath
[06287f9] Minor touchups in views
[c2400d3] Add abstract entity and attend to PR comments
[79521fc] [FIX] medical: Load XML ID directly in menu
[ba015b6] [ADD] medical: Add tour
* Force XML IDs for main menus
* Add Medical tour
[c417128] Create `oca_dependencies` and add `partner-contact`
[28ebd5f] [IMP] medical: Simplify models
* Add `partner_contact_birthdate`, `partner_firstname`, and `partner_identification`, `partner_contact_gender` dependencies
* Remove attributes in partner and patient that are duplicated by dependencies
* Remove is_* attributes on partner in favor of `type`
* Act upon type to create medical entities automatically from partner if of correct type
* Change `dob` to `birthdate_date` for compatibility with `partner_contact_birthdate`
* Change `dod` to `date_death`
* Change `deceased` to `is_deceased`
* Forward port OCA 9.0
* Update patient entity type
[821d282] [IMP] medical: Add web icon
* Add web icon to medical menu
[5123275] Remove use_parent_address
[10403f4] [MIG] medical: Upgrade to v10
* Bump version
* Change openerp namespace to odoo
* Fix inverse on res_partner to patient
[6047f0c] [FIX] medical: Patient form view
* Add terminology to archive button to fix
* Switch action_invalidate to conform with built-in toggle_active instead
* Move archive button to buttons box
* Remove configurator section, which was breaking button

jcdrubay commented on medical/README.rst:
```diff
@@ -6,83 +6,43 @@
-
-- Open Source : Licensed under AGPL
+This module extends Odoo with the base functionality of medical patients and centers.
 
```
Should you add a section"

Installation
========
You will need https://github.com/OCA/partner-contact

jcdrubay replied:
I realized that it is in the oca_dependancies.txt, but it might still worth mentioning.

jcdrubay commented on medical/models/medical_patient.py:
```diff
@@ -1,105 +1,117 @@
+    general_info = fields.Text(
+        string='General Information',
+    )
+    active = fields.Boolean(
```
Could you add a comment to explain why redefining the active field on the medical.patient? (Because the field is already available from inheritance of res.partner.

jcdrubay commented on medical/models/medical_patient.py:
```diff
@@ -1,105 +1,117 @@
+        ondelete='cascade',
+        index=True,
+    )
+    marital_status = fields.Selection([
```
This could be in a new module `partner_contact_marital_status` of https://github.com/OCA/partner-contact.

lasley replied:
Hmmm you may have a point here, I could see marital status being needed in other contexts not health related.

jcdrubay commented on medical/models/medical_patient.py:
```diff
@@ -1,105 +1,117 @@
+        Age computed depending of the birth date of the
         membership request
         """
         now = datetime.now()
```
Shouldn't it be `fields.datetime.now()`?

lasley replied:
Typically yes, but this is being used in the `relativedelta` a bit below this so the object was needed in this instance. I could have `fields.Datetime.from_string(fields.Datetime.now())`, but that would have been kind of clunky IMO

jcdrubay commented on medical/models/res_partner.py:
```diff
@@ -1,85 +1,40 @@
+
+    @api.model
+    def create(self, vals):
+        """ It overloads create to bind appropriate medical entity """
```
s/overloads/overrides

lasley replied:
The term overload is correct here IMO 😉  https://en.wikipedia.org/wiki/Function_overloading

jcdrubay replied:
Not sure because:
 - create is defined in th superclass is re-defined by this subclasses
 - the signature of the method remains the same when overriding

http://stackoverflow.com/questions/12893907/is-polymorphism-overloading-and-overriding-are-same-concepts

lasley replied:
Damn nice, seems you're right! Changing 😄

jcdrubay commented on medical/security/medical_security.xml:
```diff
@@ -69,4 +52,4 @@
 -->
-</openerp>
\ No newline at end of file
+</openerp>
```
Should be odoo.

jcdrubay commented on medical/views/medical_patient_view.xml:
```diff
@@ -1,250 +1,351 @@
+                    </group>
+                </group>
+                <notebook>
+                    <page string="General Information"> 
```
Add a name attribute to enable inheritance. In v10 I think you can only inherit without using xpath if you have a name attribute.

jcdrubay reviewed (commented):
I am far from being an expert reviewer, but I hope this helps.

lasley commented:
> I am far from being an expert reviewer, but I hope this helps.

@jcdrubay - thank you; absolutely everything helps. Ill respond or fix the line items in the next few days.

lasley commented:
@jcdrubay thank you for your review. Most comments attended to, others responded to.

Additionally, I have now added the Abstract Medical Entity model & views - this is now ready for final review! 

Once this is merged, I will submit the round of modules to cover layer two of the medical dependency chain (physician, medical_center, pharmacy, etc).

cc @t3ddftw

lasley commented on medical/views/medical_patient.xml:
```diff
@@ -0,0 +1,112 @@
+                <group name="main">
+                    <field name="identification_code" />
+                    <field name="gender" />
+                    <field name="is_pregnant" />
```
Make this a boolean_button & move to right side under `age`

lasley commented on medical/security/medical_security.xml:
```diff
@@ -1,6 +1,5 @@
-<data noupdate="0">
+<odoo noupdate="0">
 
     <record id="group_medical_user" model="res.groups">
```
These groups should be presented in a selection for the `Medical` type, instead of as independent check boxes

lasley commented on medical/views/medical_abstract_entity.xml:
```diff
@@ -0,0 +1,397 @@
+                            </form>
+                        </field>
+                    </page>
+                    <page name='page_internal_notes' string="Internal Notes">
```
Make this "Identification & Notes" + add partner_identification field

lasley commented on medical/views/medical_abstract_entity.xml:
```diff
@@ -0,0 +1,397 @@
+                <templates>
+                    <t t-name="kanban-box">
+                        <div class="oe_kanban_global_click o_res_partner_kanban">
+                            <div class="o_kanban_tags_section oe_kanban_partner_categories" />
```
The tags aren't working

jcdrubay commented on medical/models/medical_abstract_entity.py:
```diff
@@ -0,0 +1,136 @@
+            )
+            if not ids:
+                continue
+            value = [(6, 0, ids)] if many else ids[0]
```
Should this be:

if many:
   value = [(6, 0, ids)]
else:
  ids.ensure_one() # Or a nice UserWarning("Unexpectingly, several %s have been found" % category_code)
  value =  ids

lasley replied:
Good call. I changed the ambiguous naming here too - declare `ids` as a recordset and literally 3 lines down I'm using it wrong 😆

lasley replied:
Oh wait I remember the reasoning behind this now - it will be possible for this to be manipulated out from underneath the record & thus break the record. 

This is because the `id_numbers` is actually in the partner, but we are performing this pass-thru in the entity. It would also be possible by just creating a `res.partner.id_number` record outside of the partner if we weren't on the entity though, just less common. 

I proposed this feature to OCA/partner-contact, so we might be able to get a bit better handling if it gets moved there. 

Consider the following as an example:

* User adds a second driver's license through partner, but leaves the other active
* User navigates to partner's associated patient record
* UserError(*)

I'll add a comment

jcdrubay commented on medical/models/medical_abstract_entity.py:
```diff
@@ -0,0 +1,136 @@
+    @api.model
+    def _create_vals(self, vals):
+        """ Overload this in child classes in order to add values. """
+        if not vals.get('image'):
```
How about adding a if here:
```
if is_allow_default_image():
    vals['image'] = self._get_default_image(vals)
```

Then:
```
    def is_allow_default_image(self):
        if any((getattr(threading.currentThread(), 'testing', False),
                self._context.get('install_mode'))):
            if not self.env.context.get('__image_create_allow'):
                return False
        return True
```

And `_get_default_image` of the abstract would simply be a "dummy" `return None`.

Then `def _get_default_image` can focus on returning an image instead of sometimes an image or sometimes a boolean saying that we should not override the image. Also, it would simplify the overriding in `medical_patient.py` by removing this:
```
        res = super(MedicalPatient, self)._get_default_image(vals)
        if not res:
            return res
```

jcdrubay commented on medical/models/res_partner.py:
```diff
@@ -1,85 +1,39 @@
+                )
+                medical_entity = model.create(vals)
+                return medical_entity.partner_id
+        except (AttributeError, KeyError):
```
I feel worried when reading this :)

I think there should be add a comment here to explain why the exception is ignored voluntarily. Maybe a warning could also be logged.

lasley replied:
Yeah this is technically breaking Python code by excepting without an output, but I figured it would litter the logs. I will add a comment and I think I can do a little bit for legibility at the sacrifice of a bit of McCabe rating

jcdrubay commented on medical/models/medical_patient.py:
```diff
@@ -1,105 +1,113 @@
-                dod = fields.Datetime.from_string(self.dod)
-                delta = relativedelta(dod, dob)
-                deceased = _(' (deceased)')
+        for record in self:
```
Should we add a control to avoid negative ages? I am not sure if it can be required in case of pre-natal death. (I hate talking about that :/ )

lasley replied:
>  (I hate talking about that :/ )

It's sad, I think I've grown numb to this after being in medical so long. Thanks for making me realize 😛 

Perinatal mortality is measured in a different way from what I understand. There can't really be an instance of you dying before you're born, because if you weren't born how did you die? I believe those dates remain the same in this instance, although I'll bounce off a few experts to make sure.

The measurement for this is typically how far along in pregnancy the mother was (or in the instance of neonatal - up to 7 days after birth IIRC). We'll probably include this all in the pediatrics module. Or maybe something specific to prenatal even.

That said, negative ages constraint 👍

jcdrubay commented on medical/views/medical_abstract_entity.xml:
```diff
@@ -0,0 +1,404 @@
+                                                   }"
+                                    />
+                                </div>
+                                <field name="website"
```
It seems inappropriate to me on a patient form except for some very special people.

jcdrubay reviewed (changes requested):
Great code quality.

A great example of super-powered inheritances styles in Odoo in a simple and understandable context: could be used for trainings. 

I just added few comments.

I also tested in Runbot.

lasley commented:
@jcdrubay - thanks for the review, most comments attended to and some replied to.

@tedsalmon - please review.

lasley commented on medical/models/medical_abstract_entity.py:
```diff
@@ -0,0 +1,159 @@
+                record.partner_id.active = False
+
+    @api.multi
+    def _compute_identification(self, field_name, category_code, many=False):
```
This needs a inverse

lasley commented:
Added patch from LasLabs#83

tedsalmon commented on medical/models/res_partner.py:
```diff
@@ -1,85 +1,53 @@
+    @api.multi
+    @api.constrains('birthdate_date')
+    def _check_birthdate_date(self):
+        """ It will not allow birthdates in the past. """
```
s/past/future

tedsalmon reviewed (approved):
LGTM minus the aforementioned and my docblock update request.

Thanks! 👍

jcdrubay commented on medical/data/ir_sequence_data.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
Shouldn't we follow v10 standard?

`<odoo noupdate="1">` 

And remove the `<data>`

jcdrubay commented on medical/models/medical_patient.py:
```diff
@@ -1,105 +1,114 @@
+        )
+        with open(img_path, 'r') as image:
+            base64_image = image.read().encode('base64')
+            return tools.image_resize_image_big(base64_image)
```
At this step we are done with `image` already. Shouldn't the `return` be outside the `with`?

jcdrubay reviewed (changes requested):
LGTM, only 2 comments.

It is quite painful to review this big change :/ hopefully we can merge soon. :)

lasley commented:
@jcdrubay - Hah yeah this one is a beast of a change. I'm glad we're finally catching up with the repo so that we don't have to look at the entire diffs again for v11 migrations.

Thanks for the review, all points have been fixed!

jcdrubay commented:
LGTM

Hopefully we can merge very soon :) Thanks @lasley for your contributions. :+1:

lasley commented:
This is affected by LasLabs#89

lasley commented:
Ok final issue fixed - the archive button was causing some hiccups. Turns out the configurator section was also duplicating `active` and causing the archive button to not work when it was fixed too. I went ahead and removed that, as it was noted in another PR as looking bad, and I agree.

I also switched the `action_invalidate` to a `toggle_active` instead so that we conform with the built-in model operation. It still toggles for the partner too when necessary.

jcdrubay commented:
I just checked the last commit which looks good.

Now, I believe we should merge and handle remaining potential issues in separate PRs...

lasley commented:
> Now, I believe we should merge and handle remaining potential issues in separate PRs...

Yup I agree, I was planning on merging tomorrow. Honestly I was going to merge it a few days ago, but we identified the issue that day 💥
